Gazelle uses @com_google_protobuf and @com_github_golang_protobuf when `vendored = "external"` is set

When running a `gazelle` target with `external = "vendored"`, some targets are being generated which reference nonexistent `@com_google_protobuf` and `@com_github_golang_protobuf` repositories.

Specifically, I'm seeing this for the generated BUILD files for `github.com/jhump/protoreflect/desc/protoparse`.

These should be `//vendor` paths.

I too am seeing this, I'm getting output like this:
```
vendor/github.com/weaveworks/common/middleware/grpc_instrumentation.go:8:2: inconsistent definition for type grpc.callInfo during import
	struct { grpc.failFast bool; grpc.headerMD metadata.MD; grpc.trailerMD metadata.MD; grpc.peer *peer.Peer; grpc.traceInfo grpc.traceInfo; grpc.maxReceiveMessageSize *int; grpc.maxSendMessageSize *int; grpc.creds credentials.PerRPCCredentials } (in "google.golang.org/grpc")
	struct { grpc.failFast bool; grpc.headerMD metadata.MD; grpc.trailerMD metadata.MD; grpc.traceInfo grpc.traceInfo } (in "github.com/weaveworks/common/httpgrpc")
```

I think this is because I have a vendored grpc library and gazelle is adding `//vendor/google.golang.org/grpc` to my dependancies, but the go_grpc_library is depending on `@org_golang_google_grpc`.  We used to be able to override the `@org_golang_google_grpc` dependancy with the old proto rules: https://github.com/bazelbuild/rules_go/issues/442  Can we do this again?

--- a/go/tools/gazelle/resolve/resolve.go
+++ b/go/tools/gazelle/resolve/resolve.go
@@ -18,6 +18,7 @@ package resolve
 import (
 	"fmt"
 	"go/build"
+	"log"
 	"path"
 	"strings"
 
@@ -85,6 +86,7 @@ func (r *Resolver) ResolveGo(imp, pkgRel string) (Label, error) {
 const (
 	wellKnownPrefix     = "google/protobuf/"
 	wellKnownGoProtoPkg = "ptypes"
+	descriptorPkg       = "protoc-gen-go/descriptor"
 )
 
 // ResolveProto resolves an import statement in a .proto file to a label
@@ -101,14 +103,6 @@ func (r *Resolver) ResolveProto(imp, pkgRel string) (Label, error) {
 		return Label{Repo: config.WellKnownTypesProtoRepo, Name: name}, nil
 	}
 
-	// Temporary hack: guess the label based on the proto file name. We assume
-	// all proto files in a directory belong to the same package, and the
-	// package name matches the directory base name. We also assume that protos
-	// in the vendor directory must refer to something else in vendor.
-	// TODO(#859): use dependency table to resolve once it exists.
-	if pkgRel == "vendor" || strings.HasPrefix(pkgRel, "vendor/") {
-		imp = path.Join("vendor", imp)
-	}
 	rel := path.Dir(imp)
 	if rel == "." {
 		rel = ""
@@ -127,12 +121,38 @@ func (r *Resolver) ResolveGoProto(imp, pkgRel string) (Label, error) {
 
 	if isWellKnown(imp) {
 		// Well Known Type
-		pkg := path.Join(wellKnownGoProtoPkg, path.Base(imp))
-		label := r.l.LibraryLabel(pkg)
-		if r.c.GoPrefix != config.WellKnownTypesGoPrefix {
-			label.Repo = config.WellKnownTypesGoProtoRepo
+		base := path.Base(imp)
+		if base == "descriptor" {
+			switch r.c.DepMode {
+			case config.ExternalMode:
+				label := r.l.LibraryLabel(descriptorPkg)
+				if r.c.GoPrefix != config.WellKnownTypesGoPrefix {
+					label.Repo = config.WellKnownTypesGoProtoRepo
+				}
+				return label, nil
+			case config.VendorMode:
+				pkg := path.Join("vendor", config.WellKnownTypesGoPrefix, descriptorPkg)
+				label := r.l.LibraryLabel(pkg)
+				return label, nil
+			default:
+				log.Panicf("unknown external mode: %v", r.c.DepMode)
+			}
 		}
-		return label, nil
+
+		switch r.c.DepMode {
+		case config.ExternalMode:
+			pkg := path.Join(wellKnownGoProtoPkg, base)
+			label := r.l.LibraryLabel(pkg)
+			if r.c.GoPrefix != config.WellKnownTypesGoPrefix {
+				label.Repo = config.WellKnownTypesGoProtoRepo
+			}
+			return label, nil
+		case config.VendorMode:
+			pkg := path.Join("vendor", config.WellKnownTypesGoPrefix, wellKnownGoProtoPkg, base)
+			return r.l.LibraryLabel(pkg), nil
+		default:
+			log.Panicf("unknown external mode: %v", r.c.DepMode)
+		}
 	}
 
 	// Temporary hack: guess the label based on the proto file name. We assume

--- a/go/tools/gazelle/resolve/resolve_test.go
+++ b/go/tools/gazelle/resolve/resolve_test.go
@@ -147,7 +147,8 @@ func TestResolveProto(t *testing.T) {
 	prefix := "example.com/repo"
 	for _, tc := range []struct {
 		desc, imp, pkgRel      string
-		mode                   config.StructureMode
+		structureMode          config.StructureMode
+		depMode                config.DependencyMode
 		wantProto, wantGoProto Label
 	}{
 		{
@@ -162,50 +163,61 @@ func TestResolveProto(t *testing.T) {
 			wantGoProto: Label{Pkg: "foo/bar", Name: config.DefaultLibName},
 		}, {
 			desc:        "vendor",
+			depMode:     config.VendorMode,
 			imp:         "foo/bar/bar.proto",
 			pkgRel:      "vendor",
-			wantProto:   Label{Pkg: "vendor/foo/bar", Name: "bar_proto"},
+			wantProto:   Label{Pkg: "foo/bar", Name: "bar_proto"},
 			wantGoProto: Label{Pkg: "vendor/foo/bar", Name: config.DefaultLibName},
 		}, {
-			desc:        "vendor sub",
-			imp:         "foo/bar/bar.proto",
-			pkgRel:      "vendor/baz",
-			wantProto:   Label{Pkg: "vendor/foo/bar", Name: "bar_proto"},
-			wantGoProto: Label{Pkg: "vendor/foo/bar", Name: config.DefaultLibName},
+			desc:          "flat sub",
+			structureMode: config.FlatMode,
+			imp:           "foo/bar/bar.proto",
+			wantProto:     Label{Name: "foo/bar/bar_proto"},
+			wantGoProto:   Label{Name: "foo/bar"},
 		}, {
-			desc:        "flat sub",
-			mode:        config.FlatMode,
-			imp:         "foo/bar/bar.proto",
-			wantProto:   Label{Name: "foo/bar/bar_proto"},
-			wantGoProto: Label{Name: "foo/bar"},
-		}, {
-			desc:        "flat vendor",
-			mode:        config.FlatMode,
-			imp:         "foo/bar/bar.proto",
-			pkgRel:      "vendor",
-			wantProto:   Label{Name: "vendor/foo/bar/bar_proto"},
-			wantGoProto: Label{Name: "vendor/foo/bar"},
+			desc:          "flat vendor",
+			structureMode: config.FlatMode,
+			depMode:       config.VendorMode,
+			imp:           "foo/bar/bar.proto",
+			pkgRel:        "vendor",
+			wantProto:     Label{Name: "foo/bar/bar_proto"},
+			wantGoProto:   Label{Name: "vendor/foo/bar"},
 		}, {
 			desc:        "well known",
 			imp:         "google/protobuf/any.proto",
 			wantProto:   Label{Repo: "com_google_protobuf", Name: "any_proto"},
 			wantGoProto: Label{Repo: "com_github_golang_protobuf", Pkg: "ptypes/any", Name: config.DefaultLibName},
 		}, {
-			desc:        "well known flat",
-			mode:        config.FlatMode,
-			imp:         "google/protobuf/any.proto",
-			wantProto:   Label{Repo: "com_google_protobuf", Name: "any_proto"},
-			wantGoProto: Label{Repo: "com_github_golang_protobuf", Name: "ptypes/any"},
+			desc:          "well known flat",
+			structureMode: config.FlatMode,
+			imp:           "google/protobuf/any.proto",
+			wantProto:     Label{Repo: "com_google_protobuf", Name: "any_proto"},
+			wantGoProto:   Label{Repo: "com_github_golang_protobuf", Name: "ptypes/any"},
 		}, {
 			desc:        "well known vendor",
+			depMode:     config.VendorMode,
 			imp:         "google/protobuf/any.proto",
-			pkgRel:      "vendor",
 			wantProto:   Label{Repo: "com_google_protobuf", Name: "any_proto"},
-			wantGoProto: Label{Repo: "com_github_golang_protobuf", Pkg: "ptypes/any", Name: config.DefaultLibName},
+			wantGoProto: Label{Pkg: "vendor/github.com/golang/protobuf/ptypes/any", Name: config.DefaultLibName},
+		}, {
+			desc:        "descriptor",
+			imp:         "google/protobuf/descriptor.proto",
+			wantProto:   Label{Repo: "com_google_protobuf", Name: "descriptor_proto"},
+			wantGoProto: Label{Repo: "com_github_golang_protobuf", Pkg: "protoc-gen-go/descriptor", Name: config.DefaultLibName},
+		}, {
+			desc:        "descriptor vendor",
+			depMode:     config.VendorMode,
+			imp:         "google/protobuf/descriptor.proto",
+			wantProto:   Label{Repo: "com_google_protobuf", Name: "descriptor_proto"},
+			wantGoProto: Label{Pkg: "vendor/github.com/golang/protobuf/protoc-gen-go/descriptor", Name: config.DefaultLibName},
 		},
 	} {
 		t.Run(tc.desc, func(t *testing.T) {
-			c := &config.Config{GoPrefix: prefix, StructureMode: tc.mode}
+			c := &config.Config{
+				GoPrefix:      prefix,
+				DepMode:       tc.depMode,
+				StructureMode: tc.structureMode,
+			}
 			l := NewLabeler(c)
 			r := NewResolver(c, l)
 

